core-bundle/src/Resources/contao/library/Contao/StringUtil.php

```diff
@@ -582,8 +583,53 @@ static function (array $matches) use ($arrData)
 			);
 		};
 
-		$evaluateExpression = static function ($strExpression) use ($arrData)
+		// Check if we can use the expression language or if legacy tokens have been used
+		$canUseExpressionLanguage = true;
+
+		foreach (array_keys($arrData) as $token)
+		{
+			if (!preg_match('/^[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*$/', $token))
+			{
+				@trigger_error('Using tokens that are not valid PHP variables has been deprecated and will no longer work in Contao 5.0. Falling back to legacy token parsing.', E_USER_DEPRECATED);
+
+				$canUseExpressionLanguage = false;
+				break;
+			}
+		}
+
+		$expressionLanguage = null;
+
+		if ($canUseExpressionLanguage)
+		{
+			$expressionLanguage = new ExpressionLanguage();
```

Comment:
Is there a reason we don't configure a service for that? Because if I want to parse (modern) simple tokens in a Contao 4 bundle, I could then inject that service instead of using `StringUtil`. Or would it be recommended to use a "normal" ExpressionLanguage instance for that?

Reply:
@m-vo wanted to extract that into a service later on but can we please finally get this merged?

Reply:
Yeah, I think we can do that in a second step without doing any harm.

Reply:
Ping ;)

Reply:
Do we want to have a `SimpleTokenParser` service or just something for the EL bit?

Reply:
Imho the parser service because who knows what other kind of services we're going to need (and it already needs the logger afaics).

Reply:
[Pong](#1756) ;-)